### ISSUE TITLE
Question - Automatically generating a typescript interface for the global functions

### ISSUE THREAD

**Issue by alexblack** [fossamagna/gas-webpack-plugin#567]
Hi again @fossamagna , I wanted to see if you had any ideas how to improve this:

Now that we can use `gas-webpack-plugin` to automatically export our functions on the `global` object, we've reduced our the amount of repetitive code a lot, thats great.

We have one last area of repetitive code: we're manually writing a typescript interface called `ServerFunctions` which has all the exported global functions on it, so that we can call them in a type-safe manner from our addon client code.

It looks like this:

```
export interface ServerFunctions {
  setActiveSheet: (sheetName: string) => void;
  getSpreadsheetProperties: () => SpreadsheetPropertiesGAS;
  markSpreadsheetInstalledOn: () => void;
  ...
}
```

Then in the client we use it like this:

```
const Server = require('gas-client').default;

type ServerAsync = {
  serverFunctions: Promisify<ServerFunctions>;
};

const server: ServerAsync = new Server({
  // this is necessary for local development but will be ignored in production
  allowedDevelopmentDomains: (origin: string) =>
    /https:\/\/.*\.googleusercontent\.com$/.test(origin),
  parentTargetOrigin: '*',
});

export default server;
```

I'd love to find a way to not have to manually add these methods to `ServerFunctions` above.  Any ideas? I was wondering: if I initially wrote the functions on an object, then I could expose them using typescript, eg:

```
expost const SheetFunctions = {
  setActiveSheet: () => {
     // do stuff
  }
}
```

```
export type ServerFunctions = SheetFunctions;
```

but then I imagine `gas-webpack-plugin` wouldn't be able to automatically export `SheetFunctions.setActiveSheet` etc.

**Comment by fossamagna** [comment]
@alexblack Hi,  I have a cli tool that [google-script-dts-generator](https://github.com/fossamagna/google-script-dts-generator).
You can generate client side `google.script.run` type definition from server side typescript using it.
But for now it can generate from only assignment expression global object (cannot generate from named export). 
Can you once use google-script-dts-generator? (In this case you need write assignment expression to global object)
I think that I will add feature to generate from named export if it seems to meet your needs.

**Comment by enuchi** [comment]
Hi @alexblack, @fossamagna, I published a prerelease of gas-client written in typescript (`npm i gas-client@1.0.0-pr.1`) that supports what you're trying to do without needing to write the server function interfaces, and also allows autocompleting server functions. It behaves like this:

![Jun-08-2021 01-44-36](https://user-images.githubusercontent.com/31550519/121129586-2cf67480-c7fb-11eb-8339-080ff46a1840.gif)

You'll need to make a couple other changes to your code -- see the same PR with react refresh: https://github.com/enuchi/React-Google-Apps-Script/pull/98

Here's how I import `gas-client` now, including importing types from the server:

```ts
import { GASClient } from 'gas-client';
import * as publicServerFunctions from '../../server';

const { serverFunctions } = new GASClient<typeof publicServerFunctions>({
  // this is necessary for local development but will be ignored in production
  allowedDevelopmentDomains: origin =>
    /https:\/\/.*\.googleusercontent\.com$/.test(origin),
});

export { serverFunctions };
```

**Comment by alexblack** [comment]
thanks @fossamagna and @enuchi those both look like great improvements. I will take a closer look! Thanks again.

**Comment by alexblack** [comment]
thanks @fossamagna I tried out that tool, I'm not sure if I did things properly, because I don't think it found any of my methods to export:

https://gist.github.com/alexblack/0ae220accdd26f9c2ee2efb8d129d790

I ran: `yarn google-script-dts-generator -s ./src/server -o ./src/client`

Where `./src/server/index.ts` looks like this:

```
export * from './ui';
export * from './sheets';
export * from './auth';
```

**Comment by fossamagna** [comment]
@alexblack Hi

> But for now it can generate from only assignment expression global object (cannot generate from named export).
> Can you once use google-script-dts-generator? (In this case you need write assignment expression to global object)

Sorry for now cannot generate from named export. Temporarily can you write assignment expressions global object ?
I think that I will add feature to generate from named export if it seems to meet your needs.
https://github.com/fossamagna/google-script-dts-generator/issues/75

**Comment by alexblack** [comment]
@fossamagna Hi, my apologies I missed that.  Yes I can probably do that.  Am I understanding correctly: I'd temporarily go back to having `global.functionName = functionName` assignments, then use google-script-dts-generator on it to see if the output is useful/helpful here?

**Comment by fossamagna** [comment]
@alexblack Yes, If google-script-dts-generator is useful/helpful for you, I will add to support  generate from named export.

**Comment by fossamagna** [comment]
@alexblack Sorry. google-script-dts-generator has a bug, and d.ts file may not be generated correctly.
Now, I am working to fix a bug and add support `named exports`.
Tomorrow, I will publish new version of google-script-dts-generator.

**Comment by alexblack** [comment]
That sounds great, I will try it out shortly.


On Wed, Jun 09, 2021 at 7:47 AM, MURAKAMI Masahiko ***@***.***

**Comment by alexblack** [comment]
> --namedExportsFiles <glob>

@fossamagna what do I put in `<glob>`?  Is it expected to work with the setup I have now?

Where ./src/server/index.ts looks like this:

```
export * from './ui';
export * from './sheets';
export * from './auth';
```

**Comment by fossamagna** [comment]
@alexblack Thank you for trying.

Please try running as follows:
```sh
yarn google-script-dts-generator -s ./src/server/ -o ./src/client/ --namedExportsFiles ./src/server/**/*.ts
```

**Comment by alexblack** [comment]
Thanks @fossamagna sorry I should have thought of that!  Ok I tried that, and its getting closer.  With that command, here is what I get:

https://gist.github.com/alexblack/d8f8d0a1ad29c662250af2f5104003ee

I see (5) of my functions there, listed as methods of the IRun interface.  These 5 functions are from ./src/server/auth.ts.  I don't see my functions from ./src/server/sheets.ts nor ./src/server/ui.ts.

I tried other variations:
1. `yarn google-script-dts-generator -s ./src/server/ -o ./src/client/ --namedExportsFiles ./src/server/sheets.ts` works, and I then get all the functions from sheets.ts in IRun
2. `yarn google-script-dts-generator -s ./src/server/ -o ./src/client/ --namedExportsFiles ./src/server/auth.ts` works, gives me the same results as what you suggested
3. `yarn google-script-dts-generator -s ./src/server/ -o ./src/client/ --namedExportsFiles ./src/server/ui.ts` fails with an error:

```
alex@Alexs-MacBook-Air sheets-addon % yarn google-script-dts-generator -s ./src/server/ -o ./src/client/ --namedExportsFiles ./src/server/ui.ts    
yarn run v1.22.5
$ /Users/alex/dev/micro/sheets-addon/node_modules/.bin/google-script-dts-generator -s ./src/server/ -o ./src/client/ --namedExportsFiles ./src/server/ui.ts
Undefined is unsupported type. it is declarated as any type.
/Users/alex/dev/micro/sheets-addon/node_modules/google-script-dts-generator/dist/index.js:230
            members.forEach((v, k) => {
                    ^

TypeError: Cannot read property 'forEach' of undefined
    at toDomType (/Users/alex/dev/micro/sheets-addon/node_modules/google-script-dts-generator/dist/index.js:230:21)
    at /Users/alex/dev/micro/sheets-addon/node_modules/google-script-dts-generator/dist/index.js:208:64
    at Map.forEach (<anonymous>)
    at toDomType (/Users/alex/dev/micro/sheets-addon/node_modules/google-script-dts-generator/dist/index.js:206:17)
    at /Users/alex/dev/micro/sheets-addon/node_modules/google-script-dts-generator/dist/index.js:208:64
    at Map.forEach (<anonymous>)
    at toDomType (/Users/alex/dev/micro/sheets-addon/node_modules/google-script-dts-generator/dist/index.js:206:17)
    at /Users/alex/dev/micro/sheets-addon/node_modules/google-script-dts-generator/dist/index.js:199:53
    at Array.map (<anonymous>)
    at toDomType (/Users/alex/dev/micro/sheets-addon/node_modules/google-script-dts-generator/dist/index.js:199:44)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by fossamagna** [comment]
@alexblack Thank you for trying.

I can reproduce an error that `TypeError: Cannot read property 'forEach' of undefined`.
I consider that it is a bug of google-script-dts-generator.
I will fix this bug.

**Comment by fossamagna** [comment]
@alexblack I published [google-script-dts-generator v1.1.1.](https://www.npmjs.com/package/google-script-dts-generator/v/1.1.1)
This release should clear the error.

**Comment by alexblack** [comment]
Thanks @fossamagna I appreciate all your help here. Yes that clears the error.

However, I haven't yet found a way to get the methods from all 3 files (sheets.ts, auth.ts and ui.ts) into the generated interface.

When I run `yarn google-script-dts-generator -s ./src/server/ -o ./src/client/ --namedExportsFiles ./src/server/**/*.ts` I only see the methods for `auth.ts` in `google.script.d.ts`

Is it intended to work with multiple files like this?

**Comment by fossamagna** [comment]
@alexblack Hi, You need to quote the wildcard of `--namedExportsFiles` option:

```sh
yarn google-script-dts-generator -s ./src/server/ -o ./src/client/ --namedExportsFiles "./src/server/**/*.ts"
```

**Comment by alexblack** [comment]
@fossamagna You're right. Ok I think I see all the functions on the interface now, I will try out using this. Thanks!

**Comment by alexblack** [comment]
@fossamagna Its working well, it generates the interface with all functions which is great, and my client code runs.  However, when I try to build it or check it using `tsc` I get an error:

```
$ yarn tsc --project ./src/client/tsconfig.json --noEmit
$ /Users/alex/dev/micro/sheets-addon/node_modules/.bin/tsc --project ./src/client/tsconfig.json --noEmit
node_modules/@types/google.script.client-side/index.d.ts:154:11 - error TS2451: Cannot redeclare block-scoped variable 'run'.

154     const run: RunnerFunctions & PublicEndpoints;
              ~~~

  src/client/google.script.d.ts:63:15
    63         const run: IRun;
                     ~~~
    'run' was also declared here.
```

Have you seen this before?

**Comment by fossamagna** [comment]
@alexblack I guess the definition is duplicated in the type definition of `google.script.client-side` and the generated type definition.
Does excluding the `google.script.client-side` type definition eliminate the error?

**Comment by alexblack** [comment]
I'll have to give that a try when I have a sec, but I assume something else would break...

**Comment by fossamagna** [comment]
@alexblack 
When use @types/google.script.client-side, with `--endpointsOnly` option. google-script-dts-generator supported generate PublicEndpoints interface only.
This feature is released as google-script-dts-generator@1.1.2

**Comment by alexblack** [comment]
@fossamagna sorry for the delay here, thanks for your ongoing updates.  This is looking good.  I hit one issue:  some of the functions in `PublicEndpoints` have `void` as a return type, when I expected a non-void return type, eg one should have been `string | undefined`.

Here is what I see when I run it:

```
$ /Users/alex/dev/micro/sheets-addon/node_modules/.bin/google-script-dts-generator -s ./src/server/ -o ./src/client/ --namedExportsFiles './src/server/**/*.ts' --endpointsOnly
Undefined is unsupported type. it is declarated as any type.
Undefined is unsupported type. it is declarated as any type.
✨  Done in 1.54s.
```

**Comment by fossamagna** [comment]
@alexblack Thanks for your feedback.
I consider that even when a return value of server-side function is non-void type, return type of client-side function is always void.  the server-side function can return a value to the client as a parameter passed to a success handler.
https://developers.google.com/apps-script/guides/html/reference/run?hl=ja#return

**Comment by alexblack** [comment]
@fossamagna Thanks for the info, ok I think I see what you're saying.  Unfortunately in this scenario I was hoping to have the return types left unchanged.  I'm using this with https://www.npmjs.com/package/gas-client which makes it so that you can call these methods and get the return value.

eg

```
// Or we can use async/await syntax:
async () => {
  try {
    const response = await serverFunctions.addSheet(sheetTitle);
    doSomething(response);
  } catch (err) {
    handleError(err);
  }
};
```

**Comment by fossamagna** [comment]
@alexblack I understood your use case. I'll consider that add to support your use case. 
ex: When specify `--non-void-return-type` option, generate return type of server-side function not void.

**Comment by fossamagna** [comment]
@alexblack Hi, I added [--nonVoidReturnType](https://github.com/fossamagna/google-script-dts-generator#--nonvoidreturntype) option to google-script-dts-generator. This option assumed that using in combination with gas-client.This feature is released as google-script-dts-generator@1.2.0
